Editor: Add success message for scheduled posts

jeremeylduvall opened the pull request:
Attempt at fixing #4799 by adding a success message whenever a post or page is successfully scheduled in the editor.

<h3>What I Did</h3>

For the actual messaging, I mimicked what we already have in place for the publish success messaging and reworded it a tiny bit for scheduled posts/pages. It will either read `Page/Post scheduled!` or `Page/Post scheduled on siteLink` depending on what is returned from `this.props.sites.getSelectedSite()`. 

To display the message, I added an if block in `onSaveSuccess`, which checks if the `post.status` is equal to `future`. If so, it calls `this.getMessage` passing in `scheduled`, which grabs the correct message.

I thought about using an alternative approach in `onSaveDraftSuccess`, which is an intermediate that gets fired prior to `onSaveSuccess` and fetches some other messaging already. In that scenario, I would've changed the function to look like this:

```
onSaveDraftSuccess: function() {
	if ( utils.isPublished( this.state.post ) ) {
		this.onSaveSuccess(
			this.getMessage( 'updated' ),
			this.getMessage( 'view' ),
			this.state.post.URL
		);
	} else if ( utils.isFutureDated( this.state.post ) ) {
		this.onSaveSuccess(
			this.getMessage( 'scheduled' )
		);
	} else {
		this.onSaveSuccess();
	}
}
```

The main issue with that approach is that it shows the success messaging even if you haven't actually scheduled the post yet (i.e. if it autosaves while you have a future date set, but haven't pressed schedule yet) because `utils.isFutureDated( this.state.post )` eagerly returns `true`.

<h3>To Test</h3>
1. Load up this branch
2. Schedule any post or page for the future

You should see the notice when scheduling a new post/page as well as if you update a post/page that's already scheduled.

<h3>GIF</h3>

![scheduled](https://cloud.githubusercontent.com/assets/7240478/14834823/ae0ff0e0-0bc3-11e6-8f6b-5d09bc1eaf16.gif)

I threw in a semicolon for good luck in `var editorMode = 'tinymce';` because ESLint was getting angry :)

aduth commented:
Hey @jeremeylduvall , thanks for putting this together and for the included explanation.

A few thoughts:
- I prefer the alternative approach you'd considered with assigning message in `onSaveDraftSuccess`. As you noted, `isFutureDated` is a bit misleading, in which case we could use `'future' === this.state.post.status` as you'd done in your final implementation. I think by the time we've reached `onSaveSuccess`, the `message` should already be determined, so it's a bit strange to see the condition there.
- Should we have a "View Post" link in the notice as we do with the published notice?
- When updating an already-scheduled post, it still displays the "Post scheduled" notice. Do we want this to show "Post updated" instead?

jeremeylduvall commented:
Thanks @aduth. I modified the PR based on your comments and moved the work into `onSaveDraftSuccess`. I also added a link to the post preview with the wording "Preview Post" or "Preview Page." Here's a look at how it behaves now:

<img width="1007" alt="preview" src="https://cloud.githubusercontent.com/assets/7240478/14901192/783a3824-0d51-11e6-87fd-2965cab98259.png">

I agree that an already-scheduled post should probably read "Scheduled Post Updated" or something similar. I'm working on finding the right way to approach that issue. Currently `this.state.post.status` doesn't really help because it returns `future` for any post that's scheduled (whether when it's being scheduled or when a scheduled post is updated). `utils.isFutureDated` would return `true` for both cases as well.

Currently, my thought is to add `this.state.isScheduled` which is set based on the `this.state.post.status`. When the post status is set to `future`, `this.state.isScheduled` would be set to `true`, and we could update the messaging as a result. I could do a bit of a workaround for autosave as well so that the update message isn't fired anytime a scheduled post is autosaved (feels annoying). Thoughts? Is there a simpler approach I'm missing?

aduth commented:
@jeremeylduvall : I had hoped that there would be a simpler way of comparing the previous post to the new post to detect the change in status. However, following the flow of logic, the problem appears to be that the [component state is updated](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/post-editor/post-editor.jsx#L490-L525) first because the [Flux state changes](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/lib/posts/post-edit-store.js#L345-L355) before [we trigger the callback](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/lib/posts/actions.js#L373-L385) to [display the notice](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/post-editor/post-editor.jsx#L640-L650). Storing whether the post is scheduled as you suggest is an option we could pursue, but this order of events may affect whether the state has already been updated by the time we want to compare the updated post status in `onSaveSuccess`. It _might_ be that we could change the `callback` to trigger before the dispatch, or add our logic to `onEditedPostChange` before state is updated, but the ramifications of these changes are unclear to me.

Editor notices have needed to be refactored for some time now, especially since we've wanted to eliminate these action callbacks (#988). Likely outside the scope of this PR, but it would be interesting to try experimenting with different approaches for trying to detect that a notice should be shown based solely on the changing state of the post. For example, `previousPost.status === 'draft' && currentPost.status === 'publish'` should show a "Published" notice, `previousPost.status === 'publish' && currentPost.status === 'publish' && currentPost.modified !== previousPost.modified` should show an "Updated" notice.

At the end of the day, seeing "Post scheduled" instead of "Post updated" for updates to a post already scheduled does not seem like an awful experience to me, and if we'd rather move forward with what you have currently and tackle these issues separately, I'd be fine to do that.

jeremeylduvall commented:
Thanks for the explanation @aduth. That makes sense. I left the notice as "Post scheduled." The last bit I'm trying to solve is to remove the "Post scheduled" message when the post is autosaving. In the original PR, the notice would fire anytime the post was autosaved, which could get annoying. My preference would be to only see that message when I manually update the post.

The easiest solution seemed to be to look at `this.state.isAutosaving` and update accordingly. This gets a bit tricky due to how we handle scheduled posts. When [we call `autosave` for a post](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/post-editor/post-editor.jsx#L546-L580), we [check if the post is published](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/lib/posts/actions.js#L183-L208), which [sets `isAutosaving` to `true`](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/lib/posts/post-edit-store.js#L357-L360). Since the post is not published yet in this instance, we run it through [`saveEdited`](https://github.com/Automattic/wp-calypso/blob/595832734d193fca6b166975c76966c481dc5325/client/lib/posts/actions.js#L296), which (unless I'm missing something) never sets `isAutosaving` to `true`.

Dispatching again to `POST_AUTOSAVE` seemed messy, and I wasn't quite sure of the ramifications honestly. I opted to operate in `post-editor.jsx` and update `this.state.isAutosaving` to `true` in `autosave` if `this.state.post.status === 'future'`. I then set it back to `false` at the end of `onSaveSuccess`.

This _works_, but it feels a bit hacky. Sorry to lean on you so heavily here.

aduth commented:
@jeremeylduvall : I've not tried this yet in my browser, but I'm curious why this would not have been a problem previously for published posts, based on the logic. Is it the case that the notice is removed for published posts after an autosave?

aduth commented on client/post-editor/post-editor.jsx:
```diff
@@ -739,6 +778,12 @@ const PostEditor = React.createClass( {
 				this.getMessage( 'view' ),
 				this.state.post.URL
 			);
+		} else if ( 'future' === this.state.post.status && this.state.isAutosaving === false ) {
```
The latter part of the condition could be simplified to `! this.state.isAutosaving`

jeremeylduvall replied:
Fixed!

aduth commented on client/post-editor/post-editor.jsx:
```diff
@@ -97,9 +97,26 @@ const messages = {
 			return i18n.translate( 'View Post' );
 		},
+		viewPreview: function() {
+			return i18n.translate( 'Preview Post' );
```
This has me thinking about upcoming custom post type support, as there's not a label to define [when registering a custom post type](https://codex.wordpress.org/Function_Reference/register_post_type#labels) for "Preview Post" ([example API response](https://public-api.wordpress.com/rest/v1.1/sites/example.wordpress.com/post-types)). So while we could hard-code these for known post types (posts, pages), we may need to fall back to the "View" label for custom post types, which makes me question whether we should distinguish these labels in the first place. Is "View Post" confusing in the context of a scheduled post?

jeremeylduvall replied:
My rationale was that "View Post" makes me think the post is already live on my site. If I scheduled a post and noticed "View Post," I would think "Oh shoot—did it publish?"

"View Preview" could be a better compromise.

aduth replied:
>"View Preview" could be a better compromise (current version).

Indeed, I like this (or even just "Preview") as a generalized label.

jeremeylduvall commented:
@aduth for already published pages/posts, we provide [an empty callback when autosaving](https://github.com/Automattic/wp-calypso/blob/fix/4799-scheduled-post/client/post-editor/post-editor.jsx#L599) so `onSaveDraftSuccess()` isn't actually called (which results in the message not getting fired). Perhaps a better solution would be to edit [the autosave actions call](https://github.com/Automattic/wp-calypso/blob/fix/4799-scheduled-post/client/lib/posts/actions.js#L183) to include a specific logic for scheduled posts on autosave. Thoughts?

aduth commented:
@jeremeylduvall To prevent the notice from showing for autosaving of scheduled posts, could we add to the no-callback condition to include scheduled posts?

Aside: looking at that autosave component logic, I wonder if we shouldn't move the `this.setState( { isSaving: true } );` call out of the `else` block so that the `Saving...` text shows for published autosave. Definitely outside the scope of this PR.

I might try to set aside some time this week (or if you'd like to explore) to see how feasible it would be to create some kind of `<EditorNotices />` component which shows appropriate notices in response to changing post state instead of as a callback to the save actions (as [described above](https://github.com/Automattic/wp-calypso/pull/5032#issuecomment-215700469)).

jeremeylduvall commented:
Just noting that I'll take a look at this tonight/tomorrow unless you beat me to it.

jeremeylduvall commented:
> To prevent the notice from showing for autosaving of scheduled posts, could we add to the no-callback condition to include scheduled posts?

We definitely could. That was my original thought, but as you mentioned above, we don't call `this.setState( { isSaving: true } );` currently so the "Saving..." dialog wouldn't appear which seems like a poor experience. This could be fixed by moving `this.setState( { isSaving: true } );` out of the if block, but we would need to add something to the callback that switches `isSaving` back to false otherwise we'll be stuck on "Saving..." This quick hack "works" but calling `this.onSaveSuccess()` may have ramifications ([could be implemented here](https://github.com/Automattic/wp-calypso/blob/972d2fd5797e6aa4e2bcf2597be646059ca2f196/client/post-editor/post-editor.jsx#L598)).

```
this.setState( { isSaving: true } );

if ( utils.isPublished( this.state.savedPost ) || utils.isPublished( this.state.post ) || this.state.post.status === 'future' ) {
	callback = function() {
		this.onSaveSuccess();
	}.bind( this );
} else {
	callback = function( error ) {
		if ( error && 'NO_CHANGE' !== error.message ) {
			this.onSaveDraftFailure( error );
		} else {
			this.onSaveDraftSuccess();
		}
	}.bind( this );
}
```

aduth commented:
>This quick hack "works" but calling `this.onSaveSuccess()` may have ramifications

Yeah, i think we should spend some more time following the logic to see why we're specifying an empty callback for published posts. If this was intentional, your proposed change might introduce some regressions, especially since you could probably go one step further to eliminate the condition altogether (we should be checking for errors, which would be refactored to look identical to the `else` block). The presence of the condition then indicates to me that there's some reason we split the callback logic.

designsimply commented:
I tested the `fix/4799-scheduled-post` branch from #5032 using https://calypso.live/post/design5279.wordpress.com/10143 on Firefox 47.0.1 on Mac OS X 10.11.5 and it worked as expected ([39s](https://cloudup.com/cUnQqIbJerZ)), however, I'm not sure what other potential regressions (mentioned above) to test for.

aduth commented:
This will most certainly require heavy revisions with the merging of #6794 having moved notice rendering to a separate component (similar to as [proposed here earlier in the conversation](https://github.com/Automattic/wp-calypso/pull/5032#issuecomment-217879910)).

jeremeylduvall commented:
Just a note, I'm planning on looking at this tomorrow. Catching up from being out last week.

jeremeylduvall commented:
With the new separate component, I ended up creating a separate pull request in #7149. Closing this one as a result.
